File: test/test_did_peer.cpp
#include <doctest/doctest.h>

#include <string>

#include "../include/authbox.hpp"

using namespace authbox::did;

TEST_CASE("did:peer - parse simple did:peer:2 with one verification key") {
    // Example did:peer:2 with Ed25519 key
    const std::string did_uri = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";

    auto elements = parse_did_peer(did_uri);
    REQUIRE(elements.is_ok());
    CHECK(elements.value().size() == 1);
    CHECK(elements.value()[0].type == PeerElementType::Verification);
}

TEST_CASE("did:peer - parse did:peer:2 with multiple keys") {
    // did:peer:2 with verification and encryption keys
    const std::string did_uri = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK.Ez6LSbysY2xFMRpGMhb7tFTLMpeuPRaqaWM1yECx2AtzE3KCc";

    auto elements = parse_did_peer(did_uri);
    REQUIRE(elements.is_ok());
    CHECK(elements.value().size() == 2);
    CHECK(elements.value()[0].type == PeerElementType::Verification);
    CHECK(elements.value()[1].type == PeerElementType::Encryption);
}

TEST_CASE("did:peer - reject non-peer method") {
    const std::string did_uri = "did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";

    auto elements = parse_did_peer(did_uri);
    CHECK(elements.is_err());
}

TEST_CASE("did:peer - reject unsupported numalgo") {
    // did:peer:0 and did:peer:1 are not supported yet
    const std::string did_uri = "did:peer:0z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";

    auto elements = parse_did_peer(did_uri);
    CHECK(elements.is_err());
}

TEST_CASE("did:peer - reject empty method-id") {
    const std::string did_uri = "did:peer:";

    auto elements = parse_did_peer(did_uri);
    CHECK(elements.is_err());
}

TEST_CASE("did:peer - decode base58btc multibase key") {
    // Valid base58btc encoded key (z prefix)
    const std::string encoded = "z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";

    auto decoded = decode_peer_key(encoded);
    REQUIRE(decoded.is_ok());
    CHECK(decoded.value().size() > 0);
}

TEST_CASE("did:peer - reject non-base58btc encoding") {
    // Multibase with different encoding (not z)
    const std::string encoded = "u6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";

    auto decoded = decode_peer_key(encoded);
    CHECK(decoded.is_err());
}

TEST_CASE("did:peer - resolve to DID document") {
    // did:peer:2 with Ed25519 verification key
    const std::string did_uri = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";

    auto doc_result = resolve_did_peer_document_json(did_uri);
    REQUIRE(doc_result.is_ok());

    // Parse the document to verify structure
    auto document = parse_document(doc_result.value());
    REQUIRE(document.is_ok());

    CHECK(std::string_view(document.value().id.data(), document.value().id.size()) == did_uri);
    CHECK(document.value().verification_methods.size() == 1);
    CHECK(document.value().authentication.size() == 1);
}

TEST_CASE("did:peer - resolve with verification and encryption keys") {
    // did:peer:2 with both verification (V) and encryption (E) keys
    const std::string did_uri = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK.Ez6LSbysY2xFMRpGMhb7tFTLMpeuPRaqaWM1yECx2AtzE3KCc";

    auto doc_result = resolve_did_peer_document_json(did_uri);
    REQUIRE(doc_result.is_ok());

    // Parse the document
    auto document = parse_document(doc_result.value());
    REQUIRE(document.is_ok());

    CHECK(document.value().verification_methods.size() == 2);
    CHECK(document.value().authentication.size() == 1);  // V key
    CHECK(document.value().key_agreement.size() == 1);   // E key
}

TEST_CASE("did:peer - integration with resolver") {
    const std::string did_uri = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";

    // Create resolver and register did:peer handler
    Resolver resolver;
    resolver.registry().register_method("peer", [](std::string_view did_uri, const ResolveOptions &) {
        return resolve_did_peer_document_json(did_uri);
    });

    // Resolve the DID
    auto resolution = resolver.resolve(did_uri);
    REQUIRE(resolution.is_ok());

    CHECK(std::string_view(resolution.value().document.id.data(), resolution.value().document.id.size()) == did_uri);
    CHECK(resolution.value().document.verification_methods.size() == 1);
}

TEST_CASE("did:peer - element type parsing") {
    const std::string did_uri = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK.Ez6LSbysY2xFMRpGMhb7tFTLMpeuPRaqaWM1yECx2AtzE3KCc.SeyJpZCI6IiNzZXJ2aWNlIiwidCI6ImRtIiwicyI6Imh0dHBzOi8vZXhhbXBsZS5jb20vZW5kcG9pbnQifQ";

    auto elements = parse_did_peer(did_uri);
    REQUIRE(elements.is_ok());
    CHECK(elements.value().size() == 3);
    CHECK(elements.value()[0].type == PeerElementType::Verification);
    CHECK(elements.value()[1].type == PeerElementType::Encryption);
    CHECK(elements.value()[2].type == PeerElementType::Service);
}

TEST_CASE("did:peer - parse handles unknown element types gracefully") {
    // Create a did:peer with an unknown element type 'X'
    const std::string did_uri = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK.XsomeUnknownElement";

    auto elements = parse_did_peer(did_uri);
    REQUIRE(elements.is_ok());
    CHECK(elements.value().size() == 2);
    CHECK(elements.value()[0].type == PeerElementType::Verification);
    CHECK(elements.value()[1].type == PeerElementType::Unknown);
}

File: src/did/tests/test_did_peer.rs
use super::super::*;

const PEER_VERIFICATION: &str = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";
const PEER_VERIFICATION_AND_ENCRYPTION: &str = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK.Ez6LSbysY2xFMRpGMhb7tFTLMpeuPRaqaWM1yECx2AtzE3KCc";

#[test]
fn did_peer_element_default_matches_cpp_value_initialized_aggregate() {
    let element = PeerElement::default();

    assert_eq!(element.type_, PeerElementType::Verification);
    assert_eq!(element.encoded_value, "");
}

#[test]
fn did_peer_parses_simple_numalgo_2_with_one_verification_key() {
    let elements = parse_did_peer(PEER_VERIFICATION).unwrap();
    assert_eq!(elements.len(), 1);
    assert_eq!(elements[0].type_, PeerElementType::Verification);
}

#[test]
fn did_peer_parses_numalgo_2_with_multiple_keys() {
    let elements = parse_did_peer(PEER_VERIFICATION_AND_ENCRYPTION).unwrap();
    assert_eq!(elements.len(), 2);
    assert_eq!(elements[0].type_, PeerElementType::Verification);
    assert_eq!(elements[1].type_, PeerElementType::Encryption);
}

#[test]
fn did_peer_rejects_wrong_method_unsupported_numalgo_and_empty_method_id() {
    assert!(parse_did_peer("did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK").is_err());
    assert!(parse_did_peer("did:peer:0z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK").is_err());
    assert!(parse_did_peer("did:peer:").is_err());
}

#[test]
fn did_peer_decodes_base58btc_multibase_key_and_rejects_other_multibase() {
    assert!(
        !decode_peer_key("z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK")
            .unwrap()
            .is_empty()
    );
    assert!(decode_peer_key("u6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK").is_err());
}

#[test]
fn did_peer_decode_empty_key_matches_cpp_error_branch() {
    let error = decode_peer_key("").unwrap_err();
    assert_eq!(error.code, DidErrorCode::InvalidKeyFormat);
    assert_eq!(error.message, "Invalid key format: empty encoded key");
}

#[test]
fn did_peer_resolves_verification_key_to_document() {
    let document =
        parse_document(&resolve_did_peer_document_json(PEER_VERIFICATION).unwrap()).unwrap();

    assert_eq!(document.id, PEER_VERIFICATION);
    assert_eq!(document.verification_methods.len(), 1);
    assert_eq!(document.authentication.len(), 1);
    assert!(document.key_agreement.is_empty());
}

#[test]
fn did_peer_resolves_verification_and_encryption_keys_to_document() {
    let document =
        parse_document(&resolve_did_peer_document_json(PEER_VERIFICATION_AND_ENCRYPTION).unwrap())
            .unwrap();

    assert_eq!(document.id, PEER_VERIFICATION_AND_ENCRYPTION);
    assert_eq!(document.verification_methods.len(), 2);
    assert_eq!(document.authentication.len(), 1);
    assert_eq!(document.key_agreement.len(), 1);
}

#[test]
fn did_peer_relationship_refs_follow_cpp_second_pass_even_for_skipped_keys() {
    let did_uri = "did:peer:2.VuNotBase58btc.Ez6LSbysY2xFMRpGMhb7tFTLMpeuPRaqaWM1yECx2AtzE3KCc";
    let doc_json = resolve_did_peer_document_json(did_uri).unwrap();

    assert!(doc_json.contains("\"id\": \"did:peer:2.VuNotBase58btc.Ez6LSbysY2xFMRpGMhb7tFTLMpeuPRaqaWM1yECx2AtzE3KCc#key-0\""));
    assert!(doc_json.contains("\"authentication\": [\n    \"did:peer:2.VuNotBase58btc.Ez6LSbysY2xFMRpGMhb7tFTLMpeuPRaqaWM1yECx2AtzE3KCc#key-0\""));
    assert!(doc_json.contains("\"keyAgreement\": [\n    \"did:peer:2.VuNotBase58btc.Ez6LSbysY2xFMRpGMhb7tFTLMpeuPRaqaWM1yECx2AtzE3KCc#key-1\""));
}

#[test]
fn did_peer_integration_with_resolver() {
    let mut resolver = Resolver::new();
    resolver
        .registry_mut()
        .register_method("peer", |did_uri, _| resolve_did_peer_document_json(did_uri));

    let resolution = resolver.resolve(PEER_VERIFICATION).unwrap();

    assert_eq!(resolution.document.id, PEER_VERIFICATION);
    assert_eq!(resolution.document.verification_methods.len(), 1);
    assert_eq!(resolution.source_url, "did:peer:inline");
}

#[test]
fn did_peer_parses_service_and_unknown_element_types() {
    let with_service = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK.Ez6LSbysY2xFMRpGMhb7tFTLMpeuPRaqaWM1yECx2AtzE3KCc.SeyJpZCI6IiNzZXJ2aWNlIiwidCI6ImRtIiwicyI6Imh0dHBzOi8vZXhhbXBsZS5jb20vZW5kcG9pbnQifQ";
    let elements = parse_did_peer(with_service).unwrap();
    assert_eq!(elements.len(), 3);
    assert_eq!(elements[0].type_, PeerElementType::Verification);
    assert_eq!(elements[1].type_, PeerElementType::Encryption);
    assert_eq!(elements[2].type_, PeerElementType::Service);

    let with_unknown =
        "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK.XsomeUnknownElement";
    let elements = parse_did_peer(with_unknown).unwrap();
    assert_eq!(elements.len(), 2);
    assert_eq!(elements[0].type_, PeerElementType::Verification);
    assert_eq!(elements[1].type_, PeerElementType::Unknown);
}

#[test]
fn did_peer_resolves_service_element_to_document_service() {
    let did_uri = "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK.SeyJpZCI6IiNzZXJ2aWNlIiwidCI6ImRtIiwicyI6Imh0dHBzOi8vZXhhbXBsZS5jb20vZW5kcG9pbnQifQ";

    let doc_json = resolve_did_peer_document_json(did_uri).unwrap();
    assert!(doc_json.contains("\"service\""));
    assert!(doc_json.contains("https://example.com/endpoint"));

    let document = parse_document(&doc_json).unwrap();
    assert_eq!(document.services.len(), 1);
    assert_eq!(document.services[0].id, format!("{did_uri}#service"));
    assert_eq!(document.services[0].type_, "dm");
    assert_eq!(
        document.services[0].service_endpoint_json,
        "\"https://example.com/endpoint\""
    );
}

#[test]
fn did_peer_resolves_full_service_endpoint_json_without_flattening() {
    let service = r##"{"id":"#agent","type":"DIDCommMessaging","serviceEndpoint":{"uri":"https://agent.example","routingKeys":["did:example:mediator#key"]}}"##;
    let did_uri = format!(
        "did:peer:2.Vz6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK.S{}",
        base64url_encode(service.as_bytes())
    );

    let document = parse_document(&resolve_did_peer_document_json(&did_uri).unwrap()).unwrap();
    assert_eq!(document.services.len(), 1);
    assert_eq!(document.services[0].id, format!("{did_uri}#agent"));
    assert_eq!(document.services[0].type_, "DIDCommMessaging");
    assert!(
        document.services[0]
            .service_endpoint_json
            .contains("routingKeys")
    );
    assert!(
        document.services[0]
            .service_endpoint_json
            .contains("agent.example")
    );
}
